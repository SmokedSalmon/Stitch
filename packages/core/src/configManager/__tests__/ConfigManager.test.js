import { fromJS } from 'immutable'

// mock validate.js as default succeed
jest.mock('../validate', () => {
  return (schema) => jest.fn((config) => true)
})
jest.mock('../../services/constants')

// mocked config for coverage, refers: /packages/core/src/configManager/schemas/*
// updateConfig should support immutable config parameter
const mockedConfig = fromJS({
  org: {
    GBGF: 'mockedGBGF',
    EIM: 'mockedEIM',
    Product: 'mockedChannel'
  },
  env: 'Dev',
  routerPath: 'mockedRouterPath',
  hosts: [
    {
      name: 'mockedHost',
      protocol: 'https',
      server: 'localhost',
      port: 3002,
      publicPath: '/mocked-path'
    },
    { // cover branches
      name: 'mockedHostWithoutPort',
      protocol: 'https',
      server: 'localhost',
      publicPath: '/mocked-path'
    }
  ],
  libs: [{
    name: 'mockedLib',
    hostName: 'mockedHost',
    resource: '/mocked-resource/remoteEntry.js',
    styles: [{
      name: 'mockedStyle',
      hostName: 'mockedHost',
      resource: '/mocked-resource/style.css',
      autoLoad: true
    }],
    apps: [
      {
        name: 'mockedApp',
        mode: 'Web',
        routerName: ['mockedRouterName'],
        options: {},
        styles: ['mockedStyle']
      },
      { // cover branches
        name: 'mockedAppWithEmptyRouterName',
        routerName: []
      }
    ],
    services: [{
      name: 'mockedLibService',
      autoLoad: true,
      disabled: false,
      protected: true,
      options: {}
    }]
  }],
  services: [{
    name: 'mockedService',
    disabled: false,
    protected: true,
    options: {}
  }],
  globalOptions: {
    mockedOptionKey: 'mockedOptionValue'
  }
}).toJS()

let configManager = require('../index').default
let logFatal
let logError

const updateConfigErrorCode = 'CM-P-5002'

describe('ConfigManager', () => {
  beforeEach(() => {
    jest.resetModules() // reset modules for mock validate.js
  })

  test('frozen constructor', () => {
    try {
      configManager.testProperty = 'test';
    } catch (error) {
      expect(error.toString()).toBe('TypeError: Cannot add property testProperty, object is not extensible')
    }
  })

  test('validateConfig', () => {
    const { default: ConfigManager, validate } = require('../ConfigManager')
    const valid = ConfigManager.validateConfig(mockedConfig)
    const mockedValidate = validate

    expect(valid).toBe(true)
    expect(mockedValidate.mock.calls.length).toBe(1)
    expect(mockedValidate.mock.calls[0][0]).toEqual(mockedConfig)
  })

  test('validateConfig validate return false', () => {
    jest.doMock('../validate', () => {
      return (schema) => jest.fn((config) => false)
    })

    const { default: ConfigManager } = require('../ConfigManager')
    const valid = ConfigManager.validateConfig(mockedConfig)

    expect(valid).toBe(false)
  })

  test('validateConfig validate throw error', () => {
    const mockedError = new TypeError('mocked type error')

    jest.doMock('../validate', () => {
      return (schema) => jest.fn((config) => {
        throw mockedError
      })
    })
    jest.doMock('../../utils/log/Logger')

    const { default: ConfigManager } = require('../ConfigManager')
    const { default: Logger } = require('../../utils/log/Logger')

    logFatal = Logger.mock.instances[0].fatal

    try {
      ConfigManager.validateConfig(mockedConfig)
    } catch (error) {
      expect(error).toBe(mockedError)
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(mockedError)
      expect(logFatal.mock.calls[0][1]).toBe('CM-P-5001')
    }
  })

  test('updateConfig', () => {
    // can support empty object
    configManager.updateConfig({})

    configManager.updateConfig(mockedConfig)
  })

  test('getRawConfig', () => {
    expect(configManager.getRawConfig()).toMatchSnapshot()
  })

  test('getOrgConfig', () => {
    expect(configManager.getOrgConfig()).toEqual(mockedConfig.org)
  })

  test('getEnv', () => {
    expect(configManager.getEnv()).toBe(mockedConfig.env)
  })

  test('getRouterPath', () => {
    expect(configManager.getRouterPath()).toBe(mockedConfig.routerPath)
  })

  test('getGlobalOptions', () => {
    expect(configManager.getGlobalOptions()).toEqual(mockedConfig.globalOptions)
  })

  test('setGlobalOptions', () => {
    const newGlobalOptions = {
      mockedOptionKey: 'mockedOptionValueUpdated',
      mockedOptionKey2: 'mockedOptionValue2',
    }

    configManager.setGlobalOptions(newGlobalOptions)

    expect(configManager.getGlobalOptions()).toEqual(newGlobalOptions)
  })

  test('getAppConfig', () => {
    expect(configManager.getAppConfig('mockedApp')).toMatchSnapshot()
    expect(configManager.getAppConfig()).toMatchSnapshot()
  })

  test('setAppOptions', () => {
    const newAppOptions = {
      mockedAppOptionKey: 'mockedAppOptionValue'
    }

    configManager.setAppOptions('mockedApp', newAppOptions)

    expect(configManager.getAppConfig('mockedApp').options).toEqual(newAppOptions)
  })

  test('getAppName', () => {
    expect(configManager.getAppName('mockedRouterName')).toBe('mockedApp')
    expect(configManager.getAppName('mockedAppWithEmptyRouterName')).toBe('mockedAppWithEmptyRouterName')
    // cover branches
    expect(configManager.getAppName('nonexistentRouterName')).toBe('')
    expect(configManager.getAppName('')).toBe('')
  })

  test('getServiceConfig', () => {
    expect(configManager.getServiceConfig('mockedService')).toMatchSnapshot()
    expect(configManager.getServiceConfig('mockedLibService')).toMatchSnapshot()
    expect(configManager.getServiceConfig()).toMatchSnapshot()
  })

  test('setServiceOptions', () => {
    const newServiceOptions = {
      mockedServiceOptionKey: 'mockedServiceOptionValue'
    }

    configManager.setServiceOptions('mockedService', newServiceOptions)
    configManager.setServiceOptions('mockedLibService', newServiceOptions)

    expect(configManager.getServiceConfig('mockedService').options).toEqual(newServiceOptions)
    expect(configManager.getServiceConfig('mockedLibService').options).toEqual(newServiceOptions)
  })

  test('getStyleConfig', () => {
    expect(configManager.getStyleConfig('mockedLib')).toMatchSnapshot()
    expect(configManager.getStyleConfig('mockedLib', 'mockedStyle')).toMatchSnapshot()
    expect(configManager.getStyleConfig('stitch_mfe_style_1')).toMatchSnapshot()
  })
})

describe('ConfigManager.updateConfig #convertConfig failed', () => {
  beforeEach(() => {
    jest.resetModules() // reset modules for mock validate.js

    jest.doMock('../validate', () => {
      return (schema) => jest.fn((config) => true)
    })
    jest.doMock('../../utils/log/Logger')

    // require newly ConfigManager singleton instance for each test
    const { default: mockedConfigManager } = require('../index')
    configManager = mockedConfigManager

    // require mocked Logger for each test
    const { default: Logger } = require('../../utils/log/Logger')
    logFatal = Logger.mock.instances[0].fatal
  })

  test('duplicate "name" of "hosts" item', () => {
    try {
      configManager.updateConfig({
        ...mockedConfig,
        hosts: [
          ...mockedConfig.hosts,
          { name: 'mockedHost' }
        ]
      })
    } catch (error) {
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(error)
      expect(logFatal.mock.calls[0][0]).toMatchSnapshot()
      expect(logFatal.mock.calls[0][1]).toBe(updateConfigErrorCode)
    }
  })

  test('duplicate "name" of "libs" item', () => {
    try {
      configManager.updateConfig({
        ...mockedConfig,
        libs: [
          ...mockedConfig.libs,
          { name: 'mockedLib' }
        ]
      })
    } catch (error) {
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(error)
      expect(logFatal.mock.calls[0][0]).toMatchSnapshot()
      expect(logFatal.mock.calls[0][1]).toBe(updateConfigErrorCode)
    }
  })

  test('missing "hosts" item definition for "libs" item', () => {
    try {
      configManager.updateConfig({
        ...mockedConfig,
        hosts: []
      })
    } catch (error) {
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(error)
      expect(logFatal.mock.calls[0][0]).toMatchSnapshot()
      expect(logFatal.mock.calls[0][1]).toBe(updateConfigErrorCode)
    }
  })

  test('missing the matched "hosts" item', () => {
    try {
      configManager.updateConfig({
        ...mockedConfig,
        libs: [{
          ...mockedConfig.libs[0],
          styles: [{
            ...mockedConfig.libs[0].styles[0],
            hostName: 'wrongHostName'
          }]
        }]
      })
    } catch (error) {
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(error)
      expect(logFatal.mock.calls[0][0]).toMatchSnapshot()
      expect(logFatal.mock.calls[0][1]).toBe(updateConfigErrorCode)
    }
  })

  test('duplicate "name" of "styles" item', () => {
    try {
      configManager.updateConfig({
        ...mockedConfig,
        libs: [{
          ...mockedConfig.libs[0],
          styles: [
            ...mockedConfig.libs[0].styles,
            { name: 'mockedStyle' }
          ]
        }]
      })
    } catch (error) {
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(error)
      expect(logFatal.mock.calls[0][0]).toMatchSnapshot()
      expect(logFatal.mock.calls[0][1]).toBe(updateConfigErrorCode)
    }
  })

  test('missing "styles" item definition for "apps" item', () => {
    try {
      configManager.updateConfig({
        ...mockedConfig,
        libs: [{
          ...mockedConfig.libs[0],
          apps: [{
            ...mockedConfig.libs[0].apps[0],
            styles: ['wrongStyleName']
          }]
        }]
      })
    } catch (error) {
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(error)
      expect(logFatal.mock.calls[0][0]).toMatchSnapshot()
      expect(logFatal.mock.calls[0][1]).toBe(updateConfigErrorCode)
    }
  })

  test('duplicate "name" of "apps" item', () => {
    try {
      configManager.updateConfig({
        ...mockedConfig,
        libs: [{
          ...mockedConfig.libs[0],
          apps: [
            ...mockedConfig.libs[0].apps,
            { name: 'mockedApp' }
          ]
        }]
      })
    } catch (error) {
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(error)
      expect(logFatal.mock.calls[0][0]).toMatchSnapshot()
      expect(logFatal.mock.calls[0][1]).toBe(updateConfigErrorCode)
    }
  })

  test('duplicate "routerName" of "apps" item', () => {
    try {
      configManager.updateConfig({
        ...mockedConfig,
        libs: [{
          ...mockedConfig.libs[0],
          apps: [
            ...mockedConfig.libs[0].apps,
            {
              name: 'mockedApp2',
              routerName: ['mockedRouterName']
            }
          ]
        }]
      })
    } catch (error) {
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(error)
      expect(logFatal.mock.calls[0][0]).toMatchSnapshot()
      expect(logFatal.mock.calls[0][1]).toBe(updateConfigErrorCode)
    }
  })

  test('duplicate "name" of "services" item', () => {
    try {
      configManager.updateConfig({
        ...mockedConfig,
        services: [
          ...mockedConfig.services,
          { name: 'mockedService' }
        ]
      })
    } catch (error) {
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(error)
      expect(logFatal.mock.calls[0][0]).toMatchSnapshot()
      expect(logFatal.mock.calls[0][1]).toBe(updateConfigErrorCode)
    }
  })

  test('duplicate "name" of "services" item of "libs" item', () => {
    try {
      configManager.updateConfig({
        ...mockedConfig,
        libs: [{
          ...mockedConfig.libs[0],
          services: [
            ...mockedConfig.libs[0].services,
            { name: 'mockedLibService' }
          ]
        }]
      })
    } catch (error) {
      expect(logFatal.mock.calls.length).toBe(1)
      expect(logFatal.mock.calls[0][0]).toBe(error)
      expect(logFatal.mock.calls[0][0]).toMatchSnapshot()
      expect(logFatal.mock.calls[0][1]).toBe(updateConfigErrorCode)
    }
  })
})

describe('ConfigManager when globalState stitchStart = true', () => {
  beforeEach(() => {
    jest.resetModules() // reset modules for mock globalState.js

    jest.doMock('../../utils/globalState.js', () => ({
      stitchStart: true
    }))
    jest.doMock('../../utils/log/Logger')

    // require newly ConfigManager singleton instance for each test
    const { default: mockedConfigManager } = require('../index')
    configManager = mockedConfigManager

    // require mocked Logger for each test
    const { default: Logger } = require('../../utils/log/Logger')
    logError = Logger.mock.instances[0].error

    configManager.updateConfig(mockedConfig)
  })

  test('setGlobalOptions', () => {
    configManager.setGlobalOptions()

    expect(logError.mock.calls.length).toBe(1)
    expect(logError.mock.calls[0][0]).toMatchSnapshot()
    expect(logError.mock.calls[0][1]).toBe('CM-B-4001')
  })

  test('updateConfig', () => {
    configManager.updateConfig()

    expect(logError.mock.calls.length).toBe(1)
    expect(logError.mock.calls[0][0]).toMatchSnapshot()
    expect(logError.mock.calls[0][1]).toBe('CM-B-4002')
  })

  test('setAppOptions', () => {
    configManager.setAppOptions('mockedApp')

    expect(logError.mock.calls.length).toBe(1)
    expect(logError.mock.calls[0][0]).toMatchSnapshot()
    expect(logError.mock.calls[0][1]).toBe('CM-B-4003')
  })

  test('setServiceOptions', () => {
    configManager.setServiceOptions('mockedService')

    expect(logError.mock.calls.length).toBe(1)
    expect(logError.mock.calls[0][0]).toMatchSnapshot()
    expect(logError.mock.calls[0][1]).toBe('CM-B-4004')
  })
})

describe('ConfigManager cover branches', () => {
  beforeEach(() => {
    jest.resetModules() // reset modules for mock globalState.js

    // require newly ConfigManager singleton instance for each test
    const { default: mockedConfigManager } = require('../index')
    configManager = mockedConfigManager
  })

  test('updateConfig', () => {
    configManager.updateConfig({
      libs: [{
        name: 'mockedLib',
        // cover `libs` item without `apps` property
        styles: [{
          name: 'mockedStyle',
          // cover `resource` full path of `styles` item
          resource: 'http://mocked/style.css'
        }]
      }]
    })
  })
})
