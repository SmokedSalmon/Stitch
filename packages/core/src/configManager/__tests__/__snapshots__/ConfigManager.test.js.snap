// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigManager getAppConfig 1`] = `
Object {
  "libName": "mockedLib",
  "libUrl": "https://localhost:3002/mocked-path/mocked-resource/remoteEntry.js",
  "mode": "Web",
  "name": "mockedApp",
  "options": Object {},
  "routerName": Array [
    "mockedRouterName",
  ],
  "styles": Array [
    Object {
      "autoLoad": true,
      "libName": "mockedLib",
      "name": "mockedStyle",
      "styleUrl": "https://localhost:3002/mocked-path/mocked-resource/style.css",
      "uniqueID": "stitch_mfe_style_1",
    },
  ],
}
`;

exports[`ConfigManager getAppConfig 2`] = `
Object {
  "mockedApp": Object {
    "libName": "mockedLib",
    "libUrl": "https://localhost:3002/mocked-path/mocked-resource/remoteEntry.js",
    "mode": "Web",
    "name": "mockedApp",
    "options": Object {},
    "routerName": Array [
      "mockedRouterName",
    ],
    "styles": Array [
      Object {
        "autoLoad": true,
        "libName": "mockedLib",
        "name": "mockedStyle",
        "styleUrl": "https://localhost:3002/mocked-path/mocked-resource/style.css",
        "uniqueID": "stitch_mfe_style_1",
      },
    ],
  },
  "mockedAppWithEmptyRouterName": Object {
    "libName": "mockedLib",
    "libUrl": "https://localhost:3002/mocked-path/mocked-resource/remoteEntry.js",
    "mode": undefined,
    "name": "mockedAppWithEmptyRouterName",
    "options": Object {},
    "routerName": Array [
      "mockedAppWithEmptyRouterName",
    ],
    "styles": Array [],
  },
}
`;

exports[`ConfigManager getRawConfig 1`] = `
Object {
  "env": "Dev",
  "globalOptions": Object {
    "mockedOptionKey": "mockedOptionValue",
  },
  "hosts": Array [
    Object {
      "name": "mockedHost",
      "port": 3002,
      "protocol": "https",
      "publicPath": "/mocked-path",
      "server": "localhost",
    },
    Object {
      "name": "mockedHostWithoutPort",
      "protocol": "https",
      "publicPath": "/mocked-path",
      "server": "localhost",
    },
  ],
  "libs": Array [
    Object {
      "apps": Array [
        Object {
          "mode": "Web",
          "name": "mockedApp",
          "options": Object {},
          "routerName": Array [
            "mockedRouterName",
          ],
          "styles": Array [
            "mockedStyle",
          ],
        },
        Object {
          "name": "mockedAppWithEmptyRouterName",
          "routerName": Array [
            "mockedAppWithEmptyRouterName",
          ],
        },
      ],
      "hostName": "mockedHost",
      "name": "mockedLib",
      "resource": "/mocked-resource/remoteEntry.js",
      "services": Array [
        Object {
          "autoLoad": true,
          "disabled": false,
          "name": "mockedLibService",
          "options": Object {},
          "protected": true,
        },
      ],
      "styles": Array [
        Object {
          "autoLoad": true,
          "hostName": "mockedHost",
          "name": "mockedStyle",
          "resource": "/mocked-resource/style.css",
        },
      ],
    },
  ],
  "org": Object {
    "EIM": "mockedEIM",
    "GBGF": "mockedGBGF",
    "Product": "mockedChannel",
  },
  "routerPath": "mockedRouterPath",
  "services": Array [
    Object {
      "disabled": false,
      "name": "mockedService",
      "options": Object {},
      "protected": true,
    },
  ],
}
`;

exports[`ConfigManager getServiceConfig 1`] = `
Object {
  "disabled": false,
  "name": "mockedService",
  "options": Object {},
  "protected": true,
  "type": "CUSTOMIZED",
}
`;

exports[`ConfigManager getServiceConfig 2`] = `
Object {
  "autoLoad": true,
  "disabled": false,
  "libName": "mockedLib",
  "libUrl": "https://localhost:3002/mocked-path/mocked-resource/remoteEntry.js",
  "name": "mockedLibService",
  "options": Object {},
  "protected": true,
  "type": "LIB",
}
`;

exports[`ConfigManager getServiceConfig 3`] = `
Object {
  "a1": Object {
    "disabled": false,
    "name": "a1",
    "options": Object {},
    "protected": true,
    "type": "SYSTEM",
  },
  "a2": Object {
    "disabled": false,
    "name": "a2",
    "options": Object {},
    "protected": true,
    "type": "SYSTEM",
  },
  "b1": Object {
    "disabled": false,
    "name": "b1",
    "options": Object {},
    "protected": true,
    "type": "SYSTEM",
  },
  "b2": Object {
    "disabled": false,
    "name": "b2",
    "options": Object {},
    "protected": true,
    "type": "SYSTEM",
  },
  "b3": Object {
    "disabled": false,
    "name": "b3",
    "options": Object {},
    "protected": true,
    "type": "SYSTEM",
  },
  "c1": Object {
    "disabled": false,
    "name": "c1",
    "options": Object {},
    "protected": true,
    "type": "SYSTEM",
  },
  "c2": Object {
    "disabled": false,
    "name": "c2",
    "options": Object {},
    "protected": true,
    "type": "SYSTEM",
  },
  "circular_b4": Object {
    "disabled": false,
    "name": "circular_b4",
    "options": Object {},
    "protected": true,
    "type": "SYSTEM",
  },
  "d1": Object {
    "disabled": false,
    "name": "d1",
    "options": Object {},
    "protected": true,
    "type": "SYSTEM",
  },
  "d2": Object {
    "disabled": false,
    "name": "d2",
    "options": Object {},
    "protected": true,
    "type": "SYSTEM",
  },
  "mockedLibService": Object {
    "autoLoad": true,
    "disabled": false,
    "libName": "mockedLib",
    "libUrl": "https://localhost:3002/mocked-path/mocked-resource/remoteEntry.js",
    "name": "mockedLibService",
    "options": Object {},
    "protected": true,
    "type": "LIB",
  },
  "mockedService": Object {
    "disabled": false,
    "name": "mockedService",
    "options": Object {},
    "protected": true,
    "type": "CUSTOMIZED",
  },
}
`;

exports[`ConfigManager getStyleConfig 1`] = `
Array [
  Object {
    "autoLoad": true,
    "libName": "mockedLib",
    "name": "mockedStyle",
    "styleUrl": "https://localhost:3002/mocked-path/mocked-resource/style.css",
    "uniqueID": "stitch_mfe_style_1",
  },
]
`;

exports[`ConfigManager getStyleConfig 2`] = `
Array [
  Object {
    "autoLoad": true,
    "libName": "mockedLib",
    "name": "mockedStyle",
    "styleUrl": "https://localhost:3002/mocked-path/mocked-resource/style.css",
    "uniqueID": "stitch_mfe_style_1",
  },
]
`;

exports[`ConfigManager getStyleConfig 3`] = `
Array [
  Object {
    "autoLoad": true,
    "libName": "mockedLib",
    "name": "mockedStyle",
    "styleUrl": "https://localhost:3002/mocked-path/mocked-resource/style.css",
    "uniqueID": "stitch_mfe_style_1",
  },
]
`;

exports[`ConfigManager when globalState stitchStart = true setAppOptions 1`] = `"The Stitch has been started, can not set app config options anymore."`;

exports[`ConfigManager when globalState stitchStart = true setGlobalOptions 1`] = `"The Stitch has been started, can not set global config options anymore."`;

exports[`ConfigManager when globalState stitchStart = true setServiceOptions 1`] = `"The Stitch has been started, can not set service config options anymore."`;

exports[`ConfigManager when globalState stitchStart = true updateConfig 1`] = `"The Stitch has been started, can not update config anymore."`;

exports[`ConfigManager.updateConfig #convertConfig failed duplicate "name" of "apps" item 1`] = `[TypeError: Duplicate definition of apps.name (mockedApp) in config.libs]`;

exports[`ConfigManager.updateConfig #convertConfig failed duplicate "name" of "hosts" item 1`] = `[TypeError: Duplicate definition of 'name' (mockedHost) in config.hosts]`;

exports[`ConfigManager.updateConfig #convertConfig failed duplicate "name" of "libs" item 1`] = `[TypeError: Duplicate definition of 'name' (mockedLib) in config.libs]`;

exports[`ConfigManager.updateConfig #convertConfig failed duplicate "name" of "services" item 1`] = `[TypeError: Duplicate definition of 'name' (mockedService) in config.services]`;

exports[`ConfigManager.updateConfig #convertConfig failed duplicate "name" of "services" item of "libs" item 1`] = `[TypeError: Duplicate definition of 'name' (mockedLibService) in 'services' item of config.libs (mockedLib)]`;

exports[`ConfigManager.updateConfig #convertConfig failed duplicate "name" of "styles" item 1`] = `[TypeError: Duplicate definition of styles.name (mockedStyle) in config.libs (mockedLib)]`;

exports[`ConfigManager.updateConfig #convertConfig failed duplicate "routerName" of "apps" item 1`] = `[TypeError: Duplicate definition of apps.routerName[0] (mockedRouterName) in config.libs (appName: 'mockedApp2')]`;

exports[`ConfigManager.updateConfig #convertConfig failed missing "hosts" item definition for "libs" item 1`] = `[TypeError: Missing host definition of 'name' (mockedHost) in config.hosts for getting libUrl by relative path of libs resource (mockedLib)]`;

exports[`ConfigManager.updateConfig #convertConfig failed missing "styles" item definition for "apps" item 1`] = `[TypeError: Missing the definition of style (name: wrongStyleName) in config.libs (mockedLib)]`;

exports[`ConfigManager.updateConfig #convertConfig failed missing the matched "hosts" item 1`] = `[TypeError: Missing host definition of 'name' (wrongHostName) in config.hosts for getting styleUrl by relative path of libs.styles resource (mockedLib)]`;
